src/main/java/com/google/api/codegen/discovery/transformer/csharp/CSharpSampleTypeNameConverter.java

```diff
@@ -71,19 +75,47 @@ public TypeName getServiceTypeName(String apiTypeName) {
 
   @Override
   public TypeName getRequestTypeName(String apiTypeName, TypeInfo typeInfo) {
-    return getTypeName(typeInfo);
+    List<String> copy = new ArrayList<>(methodNameComponents);
+    String requestTypeName = copy.remove(copy.size() - 1) + "Request";
+    String path = "";
+    for (String s : copy) {
+      path += s + "Resource.";
+    }
+    return new TypeName(path + requestTypeName);
   }
 
   @Override
   public TypeName getTypeName(TypeInfo typeInfo) {
-    if (typeInfo.isMessage()) {
-      return new TypeName(typeInfo.message().typeName());
+    if (typeInfo.isMap()) {
+      TypeName mapTypeName = typeNameConverter.getTypeName("System.Collections.Generic.Dictionary");
+      TypeName keyTypeName = getTypeNameForElementType(typeInfo.mapKey());
+      TypeName valueTypeName = getTypeNameForElementType(typeInfo.mapValue());
```

Comment:
Yeah...... good point, I've changed it to be able to nest maps. But still, hope we never see that in production!